### PR TITLE
Optimize STAmount canonicalize:

### PR DESCRIPTION
**Pull request by seelabs**
* detect error conditions more quickly
* use correct max value in path finding

### CODE REVIEW

File: src/ripple/protocol/impl/STAmount.cpp [hunk 1]
```diff
@@ -662,13 +662,19 @@ STAmount::canonicalize()
         // native currency amounts should always have an offset of zero
         mIsNative = true;
 
-        if (mValue == 0)
+        // log(2^64,10) ~ 19.2
+        if (mValue == 0 || mOffset < -20)
```
Comment (gregtatcam) [inline src/ripple/protocol/impl/STAmount.cpp:666]: Shouldn't it be `<= -20`?

Comment (seelabs) [reply src/ripple/protocol/impl/STAmount.cpp]: Yes, I agree, we can exit early on equal as well. Fixed in `47030d563 [fold] Increase bounds on early exit`

**Comment by ximinez** [comment]
The MAC CI failures are unrelated to these changes (I think) and are resolved by #3532.

**Comment by ximinez** [comment]
After posting my previous comment Friday, I gave in to curiosity and pushed a branch with this code on top of the #3532 branch (https://github.com/ximinez/rippled/commits/sd-stamount-canonicalize) just to confirm that it would build ok. The MAC builds were fine (https://travis-ci.com/github/ximinez/rippled/builds/176166967).

The "clang-8, release" and "windows, visual studio, debug" builds had some issues, but they succeeded on retries (several in the windows case), so I'm not too worried. One odd thing that I think we may need to keep an eye on is that the first clang-8 run had several unit test failures that looked like networking problems. They went away on the retry, so it's probably just another Travis issue.

**Review by scottschurr — APPROVED** [review]
:+1: I _only_ looked at the last commit -- adjusting CrossingLimist_test.cpp.  Those changes look reasonable to me.  I left a couple of notes that could be optionally applied, but they needn't hold up the pull request.

File: src/test/app/CrossingLimits_test.cpp [hunk 2]
```diff
@@ -506,6 +522,17 @@ class CrossingLimits_test : public beast::unit_test::suite
         // up a book with many offers. At each quality keep the number of offers
         // below the limit. However, if all the offers are consumed it would
         // create a tecOVERSIZE error.
+
+        // The featureFlowSortStrands introduces a way of tracking the total
+        // number of consumed offers; with this feature the transaction no
+        // longer fails with a tecOVERSIZE error.
+        // The implementation is allows any single step to consume at most 1000
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/CrossingLimits_test.cpp:529]: Fixed in `3350bf359 [fold] Fix typos/wrong test featureset:`

File: src/test/app/CrossingLimits_test.cpp [hunk 3]
```diff
@@ -534,6 +576,7 @@ class CrossingLimits_test : public beast::unit_test::suite
         using namespace jtx;
         auto const sa = supported_amendments();
         testAll(sa - featureFlowCross);
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/CrossingLimits_test.cpp:578]: @scottschurr That was a mistake on my part. I meant to run with the featureset you specified. Nice catch!

Fixed in `3350bf359 [fold] Fix typos/wrong test featureset:`

**Review by scottschurr — APPROVED** [review]
:+1:

**Pull request closed without merging**